contracts/ReputationMiningCycle.sol

```diff
@@ -72,6 +153,305 @@ contract ReputationMiningCycle {
+    }
+    // Save the index for tiebreak scenarios later.
+    disputeRounds[u[__ROUND__]][u[__IDX__]].provedPreviousReputationUID = previousReputationUID;
+  }
+
+
+  function submitJRH(
+    uint256 round,
+    uint256 index,
+    bytes32 jrh,
+    uint branchMask1,
+    bytes32[] siblings1,
+    uint branchMask2,
+    bytes32[] siblings2
+  ) public
+  {
+    // Require we've not submitted already.
+    require(disputeRounds[round][index].jrh == 0x0);
+
+    // Check the proofs for the JRH
+    require(checkJRHProof1(jrh, branchMask1, siblings1));
+    require(checkJRHProof2(round, index, jrh, branchMask2, siblings2));
+
+    // Store their JRH
+    disputeRounds[round][index].jrh = jrh;
+    disputeRounds[round][index].lastResponseTimestamp = now;
+    disputeRounds[round][index].challengeStepCompleted += 1;
+
+    // Set bounds for first binary search if it's going to be needed
+    disputeRounds[round][index].upperBound = disputeRounds[round][index].jrhNnodes;
+  }
+
+  function checkJRHProof1(bytes32 jrh, uint branchMask1, bytes32[] siblings1) internal returns (bool result) {
+    // Proof 1 needs to prove that they started with the current reputation root hash
+    bytes32 reputationRootHash = IColonyNetwork(colonyNetworkAddress).getReputationRootHash();
+    uint256 reputationRootHashNNodes = IColonyNetwork(colonyNetworkAddress).getReputationRootHashNNodes();
+    bytes memory jhLeafValue = new bytes(64);
+    bytes memory zero = new bytes(32);
+    assembly {
+      mstore(add(jhLeafValue, 0x20), reputationRootHash)
+      mstore(add(jhLeafValue, 0x40), reputationRootHashNNodes)
+    }
+    return verifyProof(jrh, zero, jhLeafValue, branchMask1, siblings1);
+  }
+
+  function checkJRHProof2(uint round, uint index, bytes32 jrh, uint branchMask2, bytes32[] siblings2) internal returns (bool result) {
+    // Proof 2 needs to prove that they finished with the reputation root hash they submitted, and the
+    // key is the number of updates in the reputation update log (implemented)
+    // plus the number of nodes in the last accepted update, each of which will have decayed once (not implemented)
+    // TODO: Account for decay calculations
+    uint256 nUpdates = IColonyNetwork(colonyNetworkAddress).getReputationUpdateLogLength(false);
+    bytes memory nUpdatesBytes = new bytes(32);
+    disputeRounds[round][index].jrhNnodes = nUpdates + 1;
+    bytes32 submittedHash = disputeRounds[round][index].hash;
+    uint256 submittedHashNNodes = disputeRounds[round][index].nNodes;
+    bytes memory jhLeafValue = new bytes(64);
+    assembly {
+      mstore(add(jhLeafValue, 0x20), submittedHash)
+      mstore(add(jhLeafValue, 0x40), submittedHashNNodes)
+      mstore(add(nUpdatesBytes, 0x20), nUpdates)
+    }
+    return verifyProof(jrh, nUpdatesBytes, jhLeafValue, branchMask2, siblings2);
+  }
+
   function submitNewHash(bytes32 newHash, uint256 nNodes, uint256 entry) public {
     //Check the ticket is an eligible one for them to claim
```

Comment:
I think `entry` should be renamed to `entryIndex` or `entryIdx` or something together with a description of what it means in the `submittedEntries` mapping definition. We also have to change this comment to something like: "User can submit as many reputation root hashes as their staked tokens affords them, as for each submission 10^15 tokens from their stake are needed.  Here we check the entry index submitted is allowed against that stake balance"